### PR TITLE
add mv operator to SparseTensor

### LINKED ISSUE #21266: SparseTensor multiplication with 1D vector
## 🐛 Bug

Not sure if this should be a bug or a feature, but sparse tensors can only be multiplied with a 2D dense tensors, not 1D ones.  This is inconsistent with the the ordinary tensor multiplication, as well as numpy and scipy multiplication.

## To Reproduce

>>> i = torch.LongTensor([[0, 1, 1],
                          [2, 0, 2]])
>>> v = torch.FloatTensor([3, 4, 5])
>>> A = torch.sparse.FloatTensor(i, v, torch.Size([2,3]))
>>> x = torch.ones(3)
>>> A @ x[:,None] # this works
>>> A @ x # this doesn't work, but should

## Expected behavior

For a m x n sparse tensor A and a n-length dense tensor x, A @ x should give a m-length dense tensor.

### PR DESCRIPTION
**Pull request by shihongzhi**
Fixes #21266  add mv operator to SparseTensor

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
@ezyang if you have time to review this PR. Thanks.

**Review by ssnl — COMMENTED** [review]
I would love to see a doc update saying that this does not work on cuda sparse tensors. Also, @ezyang , what was our resolution for the `torch.sparse` namespace? Should this be in there or not?

**Comment by shihongzhi** [comment]
@ezyang I think this is ready for review. Thanks

**Review by ezyang — CHANGES_REQUESTED** [review]
Hi @shihongzhi, unfortunately, writing a (naive) matrix-vector multiply by hand is not the recommended way of solving the issue. One reason it's not recommended is that in your PR, there is no implementation of the kernel in CUDA; ideally, we'd also have a CUDA implementation.

Since a matrix-vector multiply is simply a matrix-matrix multiply, where the latter matrix is `n x 1` matrix, the easiest implementation is to `unsqueeze` the vector into a matrix, call the existing matrix-matrix multiply, and then squeeze it back into a vector.

**Comment by ezyang** [comment]
> Also, @ezyang , what was our resolution for the torch.sparse namespace? Should this be in there or not?

I don't remember what our resolution was. Since @weiyangfb is not working on sparse anymore, we will probably end up relitigating this when someone else picks it up. However, since `matmul` is not in sparse namespace, this probably should not be in there either.

**Comment by shihongzhi** [comment]
@ezyang  Thank for for review. I will use matrix-matrix multiply, and also to add a CUDA implementation.

File: test/test_sparse.py [hunk 1]
```diff
@@ -1430,6 +1430,15 @@ def test_log1p(self):
             device=self.device)
         self._test_log1p_tensor(input, torch.zeros([5, 6, 0]))
 
+    @cpu_only
```
Comment (ezyang) [inline test/test_sparse.py]: With this new implementation, you can make it work for CUDA too, right?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 2]
```diff
@@ -183,6 +183,29 @@ Tensor norm_sparse(const SparseTensor& self, Scalar value) {
...
+
+  TORCH_CHECK(self.ndimension() == 2 && 
+              vec.ndimension() == 1,
+              "mv: two tensor dim should be 2 and 1, but got ",
+              "SparseTensor Dim: ", self.ndimension(), "Tensor Dim: ", vec.ndimension());
+
+  TORCH_CHECK(vec.size(-1) == self.size(-1),
+              "tow tensor's last size should be equal");
+
+  auto result = self.matmul(vec.unsqueeze(-1));
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: Don't you have to squeeze the last dimension when you're done?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 3]
```diff
@@ -183,6 +183,29 @@ Tensor norm_sparse(const SparseTensor& self, Scalar value) {
...
+  TORCH_CHECK(!self.is_cuda(), "mv: expected 'self' to be CPU tensor, but got CUDA tensor");
+  TORCH_CHECK(!vec.is_cuda(), "mv: expected 'vec' to be a CPU tensor, but got a CUDA tensor");
+
+  TORCH_CHECK(self.ndimension() == 2 && 
+              vec.ndimension() == 1,
+              "mv: two tensor dim should be 2 and 1, but got ",
+              "SparseTensor Dim: ", self.ndimension(), "Tensor Dim: ", vec.ndimension());
+
+  TORCH_CHECK(vec.size(-1) == self.size(-1),
+              "tow tensor's last size should be equal");
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: nit: two

**Comment by ezyang** [comment]
Good, this is an improvement. But you need to do more:

1. Make `mv_sparse_cpu` CPU/CUDA agnostic, by:
  a. Renaming it to `mv_sparse`
  b. Removing the device checks
2. Test both cases by removing the decorator

Also I'm pretty sure you need a squeeze at the end.

Don't forget to dismiss the "changes requested" when you want more review.

**Comment by shihongzhi** [comment]
@ezyang Thank you.
I adjust the code.
It's ready for review again.

**Comment by shihongzhi** [comment]
@ezyang I think it's ready for review again.

**Comment by ezyang** [comment]
@kurtamohler and @nikitaved, since y'all are getting more involved in sparse, do you think you could help review this?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 4]
```diff
@@ -275,6 +275,27 @@ Tensor norm_sparse(const SparseTensor& self, Scalar value) {
   return self.coalesce()._values().norm(value);
 }
 
+// --------------------------------------------------------------------
+// mv(SparseTensor, Tensor)
+// --------------------------------------------------------------------
+
+Tensor mv_sparse(const SparseTensor& self, const Tensor& vec)
+{
+  AT_ASSERT(self.is_sparse());
```
Comment (kurtamohler) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: I think an is_sparse() call will always be true for a SparseTensor type.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 5]
```diff
@@ -275,6 +275,27 @@ Tensor norm_sparse(const SparseTensor& self, Scalar value) {
...
+
+  TORCH_CHECK(self.ndimension() == 2 && 
+              vec.ndimension() == 1,
+              "mv: two tensor dim should be 2 and 1, but got ",
+              "SparseTensor Dim: ", self.ndimension(), "Tensor Dim: ", vec.ndimension());
+
+  TORCH_CHECK(vec.size(-1) == self.size(-1),
+              "two tensor's last size should be equal");
```
Comment (kurtamohler) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: This message could be more clear. Maybe something like "expected self.size(-1) == vec.size(-1)"

File: test/test_sparse.py [hunk 6]
```diff
@@ -1503,6 +1503,14 @@ def test_log1p(self):
...
         self._test_log1p_tensor(input, torch.zeros([5, 6, 0]))
 
+    def test_mv(self):
+        i = self.index_tensor([[0, 1, 1], [2, 0, 2]])
+        v = self.value_tensor([3, 4, 5])
+        x = self.sparse_tensor(i, v, torch.Size([2, 3]))
+        y = torch.ones(3, device=self.device)
+
+        self.assertEqual(self.value_tensor([3, 9]), x.matmul(y))
+
```
Comment (kurtamohler) [inline test/test_sparse.py]: We should probably have test cases that make sure your asserts are triggered correctly, with `self.assertRaisesRegex`.

File: test/test_sparse.py [hunk 7]
```diff
@@ -1503,6 +1503,14 @@ def test_log1p(self):
             device=self.device)
         self._test_log1p_tensor(input, torch.zeros([5, 6, 0]))
 
+    def test_mv(self):
+        i = self.index_tensor([[0, 1, 1], [2, 0, 2]])
+        v = self.value_tensor([3, 4, 5])
+        x = self.sparse_tensor(i, v, torch.Size([2, 3]))
+        y = torch.ones(3, device=self.device)
```
Comment (kurtamohler) [inline test/test_sparse.py]: You might want to use different values for each element, not all ones, just to be safe. In fact you could think about using larger randomly generated tensors, with multiple different test cases. Check out `test_mm()` in this file.

**Review by kurtamohler — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 4 above)
Comment (shihongzhi) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: I remove this check.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 5 above)
Comment (shihongzhi) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: done

File: test/test_sparse.py (hunk 7 above)
Comment (shihongzhi) [reply test/test_sparse.py]: done

File: test/test_sparse.py (hunk 6 above)
Comment (shihongzhi) [reply test/test_sparse.py]: Sorry, I don't use `self.assertRaisesRegex` before. So I don't how and why to use this.

**Comment by shihongzhi** [comment]
@kurtamohler could you help review the code again?  thanks.

File: test/test_sparse.py (hunk 6 above)
Comment (kurtamohler) [reply test/test_sparse.py]: `self.assertRaisesRegex` is for checking to make sure functions raise exceptions when expected. The way to use it is:
```
with self.assertRaisesRegex(<type of exception>, <regex string that matches the expected error message>):
    <call the function with arguments that will cause the expected exception>
```

The error messages you're regex-matching against are the messages you wrote in your `TORCH_CHECK` calls. There are many examples in `test_sparse.py`.

**Review by kurtamohler — APPROVED** [review]

File: test/test_sparse.py (hunk 6 above)
Comment (shihongzhi) [reply test/test_sparse.py]: Done. Thank you for your detail explain.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Looks like you still need to fix some tests

```
======================================================================
FAIL: test_mv (__main__.TestUncoalescedSparse)
----------------------------------------------------------------------
RuntimeError: expected self.size(-1) == vec.size(-1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_sparse.py", line 1539, in test_mv
    test_shape(10, 100, 10, 20)
AssertionError: "expected self.size(-1) == vec.size(-1)" does not match "expected self.size(-1) == vec.size(-1)"
```

**Comment by shihongzhi** [comment]
> Looks like you still need to fix some tests
> 
> ```
> ======================================================================
> FAIL: test_mv (__main__.TestUncoalescedSparse)
> ----------------------------------------------------------------------
> RuntimeError: expected self.size(-1) == vec.size(-1)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "test_sparse.py", line 1539, in test_mv
>     test_shape(10, 100, 10, 20)
> AssertionError: "expected self.size(-1) == vec.size(-1)" does not match "expected self.size(-1) == vec.size(-1)"
> ```

Done. Fixed this bug

**Pull request closed without merging**